File: backend/services/reports.py
"""리포트 생성·조회·공유 서비스 — 비즈니스 흐름 + 예외 변환 + 단일 commit."""

from __future__ import annotations

from fastapi import HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession

from core.config import settings
from core.exceptions import ForbiddenException, ReportNotFoundException
from crud import reports as reports_crud
from db.models import SajuReport
from engine.handlers.calculate_saju import handle_calculate_saju
from llm.tools.chart_payloads import (
    payload_palja, payload_strength, payload_ten_gods, payload_wuxing_balance,
    payload_sin_sal, payload_twelve_un_seong, payload_hap_chung, payload_dae_un,
)
from schemas.report import (
    ReportCreateRequest, ReportDetail, ReportSummary, ReportShareResponse,
)

# handle_calculate_saju가 받는 kwargs (birth_input에 섞인 name 등은 제외)
_SAJU_CALC_KEYS = (
    "birth_date", "birth_time", "gender", "calendar", "is_leap_month",
    "birth_longitude", "birth_utc_offset",
)


# ─── 직렬화 헬퍼 ──────────────────────────────────────────────────────────────

def _profile_name(birth_input: dict) -> str:
    return birth_input.get("name") or ""


def _first_headline(tabs: list[dict]) -> str:
    if tabs:
        return tabs[0].get("headline", "")
    return ""


def _build_charts(birth_input: dict) -> dict | None:
    """birth_input으로 원국을 재계산해 ToolCard용 차트 payload를 조립 (best-effort).

    DB 마이그레이션 없이 조회 시점에 계산한다. 계산 실패 시 None을 반환해
    리포트 본문(탭)은 깨지지 않게 한다. payload 키는 chat tool과 동일하다.
    """
    try:
        kwargs = {k: birth_input[k] for k in _SAJU_CALC_KEYS if k in birth_input}
        saju = handle_calculate_saju(**kwargs)
        return {
            "palja":            payload_palja(saju),
            "wuxing_balance":   payload_wuxing_balance(saju),
            "strength":         payload_strength(saju),
            "ten_gods":         payload_ten_gods(saju),
            "sin_sal":          payload_sin_sal(saju),
            "twelve_un_seong":  payload_twelve_un_seong(saju),
            "hap_chung":        payload_hap_chung(saju),
            "dae_un":           payload_dae_un(saju),
        }
    except Exception:
        return None


def _to_summary(report: SajuReport) -> ReportSummary:
    return ReportSummary(
        id=report.id,
        first_headline=_first_headline(report.tabs),
        profile_name=_profile_name(report.birth_input),
        request_topics=report.request_topics,
        created_at=report.created_at.isoformat(),
    )


def _to_detail(report: SajuReport, *, mask_birth: bool = False) -> ReportDetail:
    birth_input = dict(report.birth_input)
    if mask_birth:
        birth_input["birth_date"] = None
        birth_input["birth_time"] = None
    return ReportDetail(
        id=report.id,
        first_headline=_first_headline(report.tabs),
        profile_name=_profile_name(report.birth_input),
        request_topics=report.request_topics,
        created_at=report.created_at.isoformat(),
        birth_input=birth_input,
        language=report.language,
        tabs=report.tabs,
        year_flow=report.year_flow,
        dae_un_analysis=report.dae_un_analysis,
        charts=_build_charts(report.birth_input),
    )


# ─── 유스케이스 ───────────────────────────────────────────────────────────────

async def create_report(
    db: AsyncSession,
    user_id: int,
    req: ReportCreateRequest,
    arq,
) -> int:
    """리포트 생성 job 등록 → job_id 반환. 실제 생성은 워커가 수행."""
    from services import jobs as jobs_service

    limit = settings.report_daily_limit
    if limit is not None:
        used = await reports_crud.count_today(db, user_id)
        if used >= limit:
            raise HTTPException(
                status_code=status.HTTP_429_TOO_MANY_REQUESTS,
                detail=f"하루 리포트 생성 한도({limit}개)를 초과했습니다. 내일 다시 시도해주세요.",
            )

    job_id = await jobs_service.create_job_and_enqueue(
        db,
        user_id=user_id,
        job_type="saju_report",
        payload=req.model_dump(),
        enqueue_fn="generate_saju_report",
        arq=arq,
    )
    await db.commit()
    return job_id


async def list_reports(db: AsyncSession, user_id: int) -> list[ReportSummary]:
    rows = await reports_crud.list_reports(db, user_id)
    return [_to_summary(r) for r in rows]


async def get_report(db: AsyncSession, report_id: int, user_id: int) -> ReportDetail:
    report = await reports_crud.get_report(db, report_id)
    if not report:
        raise ReportNotFoundException(str(report_id))
    if report.user_id != user_id:
        raise ForbiddenException()
    return _to_detail(report)


async def share_report(
    db: AsyncSession,
    report_id: int,
    user_id: int,
    mask_birth: bool,
) -> ReportShareResponse:
    report = await reports_crud.get_report(db, report_id)
    if not report:
        raise ReportNotFoundException(str(report_id))
    if report.user_id != user_id:
        raise ForbiddenException()

    share = await reports_crud.create_share(db, report_id, mask_birth)
    return ReportShareResponse(
        share_token=share.share_token,
        share_url=f"{settings.web_url}/share/report/{share.share_token}",
    )


async def get_shared_report(db: AsyncSession, share_token: str) -> ReportDetail:
    share = await reports_crud.get_share_by_token(db, share_token)
    if not share:
        raise ReportNotFoundException(share_token)
    report = await reports_crud.get_report(db, share.report_id)
    if not report:
        raise ReportNotFoundException(share_token)
    return _to_detail(report, mask_birth=share.mask_birth)

File: packages/api-client/src/reports.ts
import type { ApiClient } from './client'
import type { SajuCalcRequest } from './types'
import type { JobCreated } from './jobs'

// ── 공통 API 계약 타입 (plan §트랙B — 변경 금지) ──────────────────────────────

export interface ReportTab {
  category: string          // 짧은 태그 — "성격" | "재물" | ... | 요청 주제명
  headline: string          // 결론형 문장
  content: string           // 3~4문단 (비유→근거→조언, \n\n 구분)
  requested: boolean        // 사용자가 추가 요청한 주제 탭이면 true
}

export interface YearFlowMonth {
  month: number
  keyword: string
  memo: string
}

export interface YearFlow {
  year: number
  first_half: string        // 상반기 요약 2~3문장
  second_half: string
  months: YearFlowMonth[]   // 12개
}

export interface DaeUnAnalysis {
  current: { ganji: string; period: string; text: string }   // period 예: "현재 ~2026"
  next:    { ganji: string; period: string; text: string }
  caution: string           // 주의점 1문단
}

export interface ReportSummary {
  id: number
  first_headline: string
  profile_name: string      // birth_input.name
  request_topics: string | null
  created_at: string
}

export interface ReportDetail extends ReportSummary {
  birth_input: SajuCalcRequest      // mask 시 birth_date/birth_time null
  language: string
  tabs: ReportTab[]                 // 10개 + 요청 주제 수만큼
  year_flow: YearFlow
  dae_un_analysis: DaeUnAnalysis
  charts?: Record<string, unknown>  // 원국 차트 payload (palja·wuxing_balance·strength·ten_gods), 계산 실패 시 부재
}

export interface CreateReportBody {
  birth_input: SajuCalcRequest
  request_topics?: string
  profile_id?: number
  language?: string  // 생성 언어 ('ko' | 'en'), 기본 'ko'
}

export interface ShareReportResponse {
  share_token: string
  share_url: string
}

// ── API 함수 ──────────────────────────────────────────────────────────────────

/** POST /api/reports — 리포트 생성 job 등록 (로그인 필수). 202 { job_id } */
export function createReportJob(
  api: ApiClient,
  body: CreateReportBody,
): Promise<JobCreated> {
  return api.post<JobCreated>('/api/reports', body)
}

/** GET /api/reports — 내 리포트 목록 (로그인). */
export function listReports(api: ApiClient): Promise<ReportSummary[]> {
  return api.get<ReportSummary[]>('/api/reports')
}

/** GET /api/reports/{id} — 단건 조회 (소유자만). */
export function getReport(api: ApiClient, id: number): Promise<ReportDetail> {
  return api.get<ReportDetail>(`/api/reports/${id}`)
}

/** DELETE /api/reports/{id} — 리포트 삭제 (소유자). 204 No Content */
export function deleteReport(api: ApiClient, id: number): Promise<void> {
  return api.delete(`/api/reports/${id}`)
}

/** POST /api/reports/{id}/share — 공유 토큰 발급 (소유자). */
export function shareReport(
  api: ApiClient,
  id: number,
  maskBirth: boolean,
): Promise<ShareReportResponse> {
  return api.post<ShareReportResponse>(`/api/reports/${id}/share`, { mask_birth: maskBirth })
}

/** GET /api/share/reports/{token} — 비로그인 공개 열람. */
export function getSharedReport(
  api: ApiClient,
  token: string,
): Promise<ReportDetail> {
  return api.get<ReportDetail>(`/api/share/reports/${token}`)
}
